text/0000-loop-break-value.md

```diff
@@ -0,0 +1,216 @@
+- Feature Name: loop_break_value
+- Start Date: 2016-05-20
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+(This is a result of discussion of
+[issue #961](https://github.com/rust-lang/rfcs/issues/961) and related to RFCs
+[352](https://github.com/rust-lang/rfcs/pull/352) and
+[955](https://github.com/rust-lang/rfcs/pull/955).)
+
+Let a `loop { ... }` expression return a value via `break my_value;`.
+
+# Motivation
+[motivation]: #motivation
+
+This pattern is currently hard to implement without resorting to a function or
+closure wrapping the loop:
```

Comment:
Damn. The problem with being British is you expect people to spot the sarcasm.

1.  Having to mutable a variable just to return a value is inelegant
2.  Having to wrap with `Option` then unwrap just because there isn't an appropriate default value is clumsy
3.  Having to do this while considering the lifetimes of generated references is confusing
4.  It's a lot of extra code. Compare the above to:

```rust
let computation = loop {
        if let Some(r) = self.do_something() {
            break r;
        }
    }.do_computation();
self.use(computation);
```
(noting that due to reference lifetimes it may be possible to give the value fed to `do_computation()` a name in the outer scope).

Is that sufficient motivation?

The new lifetime rules under discussion may well simplify this problem, both with-and-without break-with-value, but I still think break-with-value is a neat pattern (or alternately, a generic break-from-block-with-value, but that opens another can of worms: the value if not broken out of).